### PR TITLE
Improved code coverage for pkg/kubelet/types/pod_update

### PR DESCRIPTION
**Pull request by gyliu513**
The test coverage for pod_update.go was imprved from 36% to 100%.

**What this PR does / why we need it**:
This fixed part of #40780
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by tmrts** [comment]
@k8s-bot ok to test

**Comment by gyliu513** [comment]
@k8s-bot verify test this

**Comment by gyliu513** [comment]
@k8s-bot verify test this

**Comment by gyliu513** [comment]
@k8s-bot non-cri e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot cvm gce e2e test this

**Comment by gyliu513** [comment]
@k8s-bot gce etcd3 e2e test this
@k8s-bot cvm gce e2e test this

**Comment by tmrts** [comment]
@k8s-bot test this

it says the node test suite failed due to overload

File: pkg/kubelet/types/pod_update_test.go
Comment (tmrts) [inline pkg/kubelet/types/pod_update_test.go]: If its an error, you have to assume that it won't return a value or it will be meaningless: don't use `source` if `err != nil`

Comment (tmrts) [inline pkg/kubelet/types/pod_update_test.go]: If its an error, you have to assume that it won't return a value or it will be meaningless: don't use `source` if `err != nil`

File: pkg/kubelet/types/pod_update_test.go
Comment (tmrts) [inline pkg/kubelet/types/pod_update_test.go]: What does this case accomplish that the one above it doesn't?

**Review by tmrts — CHANGES_REQUESTED** [review]

**Comment by tmrts** [comment]
LGTM after a few comments and green tests :+1:

File: pkg/kubelet/types/pod_update_test.go
Comment (gyliu513) [inline pkg/kubelet/types/pod_update_test.go]: done

File: pkg/kubelet/types/pod_update_test.go
Comment (gyliu513) [inline pkg/kubelet/types/pod_update_test.go]: killed this, thanks @tmrts

File: pkg/kubelet/types/pod_update_test.go
Comment (gyliu513) [inline pkg/kubelet/types/pod_update_test.go]: oops, @tmrts , the problem is that from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go#L107 , the source will be `""` and also an error will be returned, otherwise, it will be expected errors, so here I was using `if err != nil && source != ""`, what do you think?

If update as 
```
func TestGetPodSource(t *testing.T) {
        cases := []struct {
                pod      v1.Pod
                expected string
        }{
                {
                        pod:      v1.Pod{},
                        expected: "",
                },
                {
                        pod: v1.Pod{
                                ObjectMeta: metav1.ObjectMeta{
                                        Annotations: map[string]string{
                                                "kubernetes.io/config.source": "host-ipc-sources",
                                        },
                                },
                        },
                        expected: "host-ipc-sources",
                },
        }
        for _, data := range cases {
                source, err := GetPodSource(&data.pod)
                if err != nil {
                        t.Errorf("unexpected error for source %v: %v", source, err)
                }
                assert.Equal(t, data.expected, source)
        }
}
```

The test will be failed as 
```
--- FAIL: TestGetPodSource (0.00s)
	pod_update_test.go:72: unexpected error for source : cannot get source of pod ""
```

**Comment by tmrts** [comment]
@k8s-bot test this

File: pkg/kubelet/types/pod_update_test.go
Comment (tmrts) [inline pkg/kubelet/types/pod_update_test.go]: @gyliu513 you don't have to do those checks, just use `assert.NoError` for the error, that should be enough

Comment (gyliu513) [inline pkg/kubelet/types/pod_update_test.go]: The problem is that this test case will actually generate an error

```
               {
                        pod:      v1.Pod{},
                        expected: "",
                },
```

Because of https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go#L107 , so seems I cannot use `assert.NoError` here

Comment (tmrts) [inline pkg/kubelet/types/pod_update_test.go]: @gyliu513 in that case, you should add a new field errExpected: true/false, and use an additional check for that one `if err != nil && !errExpected`

Comment (tmrts) [inline pkg/kubelet/types/pod_update_test.go]: or you can use if errExpected { assert.Error } else { assert.NoError}

**Comment by gyliu513** [comment]
@k8s-bot unit test this

**Comment by gyliu513** [comment]
@k8s-bot non-cri node e2e test this

**Comment by gyliu513** [comment]
@timstclair can you please help check if this can be approved? I do not know how to remove the `do-not-merge` flag, can you help? Thanks!

**Comment by tmrts** [comment]
@dchen1107 PR is ready for the final pass

**Comment by gyliu513** [comment]
Rebase, @dchen1107 @tmrts , can you please help check if this can be merged? Thanks.

**Comment by gyliu513** [comment]
@dchen1107 can you please help check if this can be merged? Thanks.

**Comment by gyliu513** [comment]
@dchen1107 rebased, can you please help check if this can be merged? Thanks.

**Comment by gyliu513** [comment]
@dchen1107 @tmrts rebased, can you please help check if this can be merged? Thanks.

**Comment by gyliu513** [comment]
@k8s-bot gce etcd3 e2e test this

File: pkg/kubelet/types/pod_update_test.go
Comment (yujuhong) [inline pkg/kubelet/types/pod_update_test.go]: You should make it clear which test case failed.
For example,
```
assert.Equal(t, data.expected, actual, "Annotations: %+v",  data.Pod.Annotations)
```

File: pkg/kubelet/types/pod_update_test.go
Comment (yujuhong) [inline pkg/kubelet/types/pod_update_test.go]: You should make it clear which test case failed. Either add a description, or at the least, print the test case number.
E.g.,
```
for i, data := range cases {
 ...
 assert.Equal(t, data.expected, source, "test[%d]", i)
```
Alternatively, log the information in the loop:
```
t.Logf("Test case [%d]", i)
```

**Review by yujuhong — CHANGES_REQUESTED** [review]

File: pkg/kubelet/types/pod_update_test.go
Comment (gyliu513) [inline pkg/kubelet/types/pod_update_test.go]: Done

File: pkg/kubelet/types/pod_update_test.go
Comment (gyliu513) [inline pkg/kubelet/types/pod_update_test.go]: Also updated here

File: pkg/kubelet/types/pod_update_test.go
Comment (gyliu513) [inline pkg/kubelet/types/pod_update_test.go]: Updated the error message by adding test case index

**Comment by gyliu513** [comment]
ping @yujuhong , can you please help approve this? All comments are now addressed. Thanks.

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by gyliu513** [comment]
This PR need love, it has been here for almost half year, and I got lgtm label, then rebase, and lost lgtm label and rebase, got lgtm again, then rebase ...

@tmrts @yujuhong can you help? If you do not have cycle, can you please help find someone to review this? Thanks!

**Review by yujuhong — APPROVED** [review]

**Pull request merged**
